<?xml version="1.0"?>

<robot name="manipulator">

    <link name="world">
        <visual>
            <origin xyz="0 0 -0.025"/>
            <geometry>
                <box size="2 2 0.05"/>
            </geometry>
            <material name="grey">
                <color rgba="1 1 1 1"/>
            </material>
        </visual>
    </link>

    <link name="l1">
        <visual>
            <origin xyz="0 0 -0.05"/>
            <geometry>
                <cylinder length="0.1" radius="0.08"/>
            </geometry>
            <material name="light_white">
                <color rgba="1 1 1 1"/>
            </material>
        </visual>
    </link>

    <link name="l2">
        <visual>
            <origin xyz="0.25 0 0" rpy="0 1.57 0" />
            <geometry>
                <cylinder length="0.5" radius="0.03" />
            </geometry>
            <material name="blue">
                <color rgba="0.3 0.6 1 1" />
            </material>
        </visual>
    </link>

    <link name="l3">
        <visual>
            <origin xyz="0 0.175 0" rpy="1.57 0 0"/>
            <geometry>
                <cylinder length="0.35" radius="0.03"/>
            </geometry>
            <material name="color_custom">
                <color rgba="1 0.4 0.9 1"/>
            </material>
        </visual>
    </link>

    <link name="l4">
        <visual>
            <origin xyz="0 0 -0.075"/>
            <geometry>
                <box size="0.07 0.07 0.15"/>
            </geometry>
            <material name="green">
                <color rgba="0.4 1 0.4 1"/>
            </material>
        </visual>
    </link>

    <link name="l5">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <cylinder length="0.1" radius="0.02"/>
            </geometry>
            <material name="orange">
                <color rgba="1 0.36 0.23 1"/>
            </material>
        </visual>
    </link>

    <link name="l6">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry> 
                <box size="0.1 0.1 0.03"/> 
            </geometry>
            <material name="white">
                <color rgba="0.2 0.2 1 1"/>
            </material>
        </visual>
    </link>

    <joint name="world_to_l1" type="revolute">
        <parent link="world"/>
        <child link="l1"/>
        <origin xyz="0 0 0.1"/>
        <axis xyz="0 0 1"/>
        <limit lower="-3.1" upper="3.1" effort="0.1" velocity="0.2"/>
    </joint>

    <joint name="l1_to_l2" type="revolute">
        <parent link="l1"/>
        <child link="l2"/>
        <origin xyz="0 0 0" rpy="-1.57 -1.57 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-1.57" upper="1.57" effort="0.1" velocity="0.2"/>
    </joint>

    <joint name="l2_to_l3" type="revolute">
        <parent link="l2"/>
        <child link="l3"/>
        <origin xyz="0.5 0 0" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-3.1" upper="3.1" effort="0.1" velocity="0.2"/>
    </joint>

    <joint name="l3_to_l4" type="revolute">
        <parent link="l3"/>
        <child link="l4"/>
        <origin xyz="0 0.5 0" rpy="-1.57 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-3.1" upper="3.1" effort="0.1" velocity="0.2"/>
    </joint>

    <joint name="l4_to_l5" type="revolute">
        <parent link="l4"/>
        <child link="l5"/>
        <origin xyz="0 0 0" rpy="1.57 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-3.1" upper="3.1" effort="0.1" velocity="0.2"/>
    </joint>

    <joint name="l5_to_l6" type="revolute">
        <parent link="l5"/>
        <child link="l6"/>
        <origin xyz="0 0 0" rpy="-1.57 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-3.1" upper="3.1" effort="0.1" velocity="0.2"/>
    </joint>

</robot>